Bug/exclusion dates timezones and month change

Erbidium opened the pull request.

Erbidium committed:
Fix month change

Erbidium committed:
Merge branch 'development' into bug/exclusion-dates-timezones-and-month-change

Erbidium committed:
Convert new exclusion date to utc, convert all exclusion dates to user selected timezone


Refactored convertExclusionDateToUtc method


Simplify method


Move functions to separate helpers


Move enum to separate file, fix helpers


Fix enum import


Fix time string helper


Move functions to separate helpers, start to create exclusion dates display


Make arrow functions


Refactor helpers, update exclusion dates after timezone change


update exclusion dates after adding new ones


Delete exclusion date


Fix special cases


Create end time range pipe


Revert "Create end time range pipe"

This reverts commit e46d340e5fe4f93d5b51dd1aba335b53c560fa1d.

Fix time conversion


lint fix


Replace for of loops

Erbidium committed:
Merge branch 'development' into bug/exclusion-dates-timezones-and-month-change

Erbidium committed:
Refactor exclusion date component


Fix timezone change update


Filter empty exlcusion dates

Erbidium committed:
Move helper to another helper


Lint fix

Erbidium committed:
Remove length check

YuraRiabov commented on frontend/src/app/core/helpers/exclusion-date-helper.ts:
```diff
@@ -0,0 +1,121 @@
+                getExclusionDate(endRangeDate, 0, '00', endHours, endMinutes),
+            ];
+        })
+        .flat();
```
Please refactor this method

Erbidium committed:
Refactor exclusion dates conversion method


Fix minutes extraction

Erbidium committed:
Fix exclusion dates deleting

tatianahutii commented on frontend/src/app/core/helpers/exclusion-date-helper.ts:
```diff
@@ -0,0 +1,150 @@
+) {
+    let convertedDate: Date;
+
+    if (dayAction === DayAction.AddDay) {
+        convertedDate = dateAdapter.addCalendarDays(new Date(exclusionDate.selectedDate), 1);
+    } else if (dayAction === DayAction.SubtractDay) {
+        convertedDate = dateAdapter.addCalendarDays(new Date(exclusionDate.selectedDate), -1);
+    } else {
+        convertedDate = new Date(exclusionDate.selectedDate);
+    }
```
Please refactor

Erbidium replied:
Fixed by using switch case

tatianahutii commented on frontend/src/app/core/helpers/exclusion-date-helper.ts:
```diff
@@ -0,0 +1,150 @@
+): [string, string] =>
+    (startHoursDayAction === DayAction.SubtractDay
+        ? [
+            getDateWithoutLocalOffset(dateAdapter.addCalendarDays(new Date(exclusionDate.selectedDate), -1)).toJSON(),
```
Why do we need JSON?

Erbidium replied:
In order to convert date to json string and then pass this string in request. Native date object during convertation to json is converted with user offset from his PC. To avoid this I convert it manually and then pass string, when I need Date object, I create it from string right in place

Erbidium replied:
Fixed it by moving toJSON() to helper method

tatianahutii commented on frontend/src/app/core/helpers/exclusion-date-helper.ts:
```diff
@@ -0,0 +1,150 @@
+                timeZoneValue,
+            );
+            const [endHours, endMinutes, endHoursDayAction] = convertTimeToOffset(
+                dayTimeRange.end === '23:59' && !timeZoneValue.includes('00:00') ? '24:00' : dayTimeRange.end,
+                timeZoneValue,
```
Looks very starnge

Erbidium replied:
The time range is stored as two strings and I manually convert them to a user time zone. Sometimes a time range can be divided into two parts: one for this day and part for the next day. I consider that day starts at 00:00 and ends at 23:59. It means that during conversion 22:00(UTC) can be converted to 23:59(+02:00), but when I convert backwards, I need 24:00 value in order to make everything work correctly

Erbidium replied:
Fixed

tatianahutii commented on frontend/src/app/core/helpers/exclusion-date-helper.ts:
```diff
@@ -0,0 +1,150 @@
+                startHoursDayAction,
+            );
+
+            if (endHours === 0 && endMinutes === '00') {
+                return [getExclusionDate(startRangeDate, startHours, startMinutes, 23, '59')];
+            }
```
see the previous comment

Erbidium replied:
Fixed

tatianahutii commented on frontend/src/app/core/helpers/exclusion-date-helper.ts:
```diff
@@ -0,0 +1,150 @@
+) {
+    let updatedExclusionDatesDisplay: IExclusionDate[] = [];
+
+    exclusionDates.forEach((exclusionDate) => {
+        const convertedExclusionDates = convertExclusionDateToOffset(exclusionDate, timeZoneValue, dateAdapter);
+
+        convertedExclusionDates.forEach((convertedExclusionDate) => {
+            if (!mergeExistingExclusionDates(convertedExclusionDate, updatedExclusionDatesDisplay)) {
+                updatedExclusionDatesDisplay = [...updatedExclusionDatesDisplay, convertedExclusionDate];
+            }
+        });
+    });
```
Please get rid of forEach in forEach

Erbidium replied:
Fixed

tatianahutii commented on frontend/src/app/core/helpers/time-helper.ts:
```diff
@@ -1,6 +1,35 @@
+        (control: AbstractControl): ValidationErrors | null =>
+            (firstControl?.value < control?.value ? null : { error: true });
+
+export function normalizeHours(hours: number): [number, DayAction] {
+    if (hours < 0) {
+        return [24 + hours, DayAction.SubtractDay];
+    }
+    if (hours >= 24) {
+        return [hours - 24, DayAction.AddDay];
+    }
+
+    return [hours, DayAction.NoAction];
+}
```
Why do we need this?

Erbidium replied:
I convert string hours to user selected time zone. After adding timeZone hours to the parsed hours value, it can become negative or more than 24. This used to give hours appropriate values

Erbidium replied:
Fixed

tatianahutii commented on frontend/src/app/core/helpers/time-helper.ts:
```diff
@@ -1,6 +1,35 @@
+    return [hours, DayAction.NoAction];
+}
+
+export const getTimeString = (hours: number, minutes: string) =>
+    `${hours.toString().length < 2 ? '0' : ''}${hours}:${minutes}`;
```
Why do we need this?

Erbidium replied:
We need this in order to have always string in format "00:00" with four digits and ':' symbol

Erbidium replied:
Fixed

tatianahutii commented on frontend/src/app/core/helpers/time-helper.ts:
```diff
@@ -1,6 +1,35 @@
+export const getTimeString = (hours: number, minutes: string) =>
+    `${hours.toString().length < 2 ? '0' : ''}${hours}:${minutes}`;
+
+export function convertTimeToOffset(time: string, offsetValue: string): [number, string, DayAction] {
+    const offsetHours = Number(offsetValue.substring(0, 3));
+    const hours = Number(time.substring(0, 2));
+    const minutes = time.substring(3, 5);
+    const hoursConverted = hours + offsetHours;
+
+    const [correctHours, hoursDayAction] = normalizeHours(hoursConverted);
+
+    return [correctHours, minutes, hoursDayAction];
```
Looks strange

Erbidium replied:
Refactored

tatianahutii commented on frontend/src/app/core/helpers/time-helper.ts:
```diff
@@ -1,6 +1,35 @@
+    return [correctHours, minutes, hoursDayAction];
 }
+
+export const changeOffsetSign = (offset: string) =>
+    (offset[0] === '+' ? offset.replace(/\+/g, '-') : offset.replace(/-/g, '+'));
```
the same. looks starnge

Erbidium replied:
It is done to get opposite value to user timeZone value: "+02:00" to "-02:00" and opposite

Erbidium replied:
Fixed

tatianahutii commented on frontend/src/app/modules/availability/new-slot/schedule-definition/exclusion-date/exclusion-date.component.ts:
```diff
@@ -1,24 +1,117 @@
+    }
+
     @Input() set schedule(value: ISchedule | undefined) {
         this.scheduleValue = value ?? getDefaultSchedule(false);
+        this.scheduleValue.exclusionDates.forEach((date) =>
+            date.dayTimeRanges.forEach((range) => {
+                range.start = this.formatTime(range.start);
+                range.end = this.formatTime(range.end);
+            }));
```
why forEach?

Erbidium replied:
Fixed(Removed)

tatianahutii commented on frontend/src/app/modules/availability/new-slot/schedule-definition/exclusion-date/exclusion-date.component.ts:
```diff
@@ -1,24 +1,117 @@
+    }
 
     deleteExclusionDate(index: number) {
+        const exclusionDateDisplayToDelete = this.exclusionDatesDisplay[index];
+        const exclusionDateDisplaysToDeleteUtc = convertExclusionDateToOffset(
+            exclusionDateDisplayToDelete,
+            changeOffsetSign(this.scheduleValue.timeZone.timeValue),
+            this.dateAdapter,
+        );
```
could we ret rid of the index?

Erbidium replied:
Fixed

tatianahutii commented on frontend/src/app/modules/availability/new-slot/schedule-definition/exclusion-date/exclusion-date.component.ts:
```diff
@@ -1,24 +1,117 @@
+                        ),
+                );
+
+                sameDate.dayTimeRanges.forEach((range) => {
+                    date.dayTimeRanges.forEach((rangeToDelete) => {
```
see the previous comment

Erbidium replied:
Fixed

tatianahutii commented on frontend/src/app/modules/availability/new-slot/schedule-definition/exclusion-date/exclusion-date.component.ts:
```diff
@@ -1,24 +1,117 @@
+            .afterClosed()
+            .subscribe((newExclusionDates) => {
+                if (newExclusionDates) {
+                    newExclusionDates.forEach((newExclusionDate) => {
```
why forEach?

Erbidium replied:
After user selects exclusion date with time ranges, according to selected timezone it can be converted to two exclusion dates in utc

Erbidium replied:
Fixed

Erbidium committed:
Merge branch 'development' into bug/exclusion-dates-timezones-and-month-change

Erbidium committed:
Move toJSON() to helper, refactor if else if


Simplify hours normalization

Erbidium committed:
Use timeonly, add JsonConverter

Erbidium committed:
Fix time formatting, fix time ranges comparison


Fix comparison

Erbidium committed:
Refactor timeZone value conversion and sign change

Erbidium committed:
Merge remote-tracking branch 'origin/development' into bug/exclusion-dates-timezones-and-month-change

Erbidium committed:
Use DateTime for exclusion dates


Rework exclusion dates


Fix exclusion date timezone conversion


Fix timeZone conversion


Add sinlge exclusion time range deleting


Add exclusion dates display


Include starting or ending dates


Filter display with empty ranges, check if range between end and date is not zero


Create pipe to format hours


Delete exclusion dates


Lint fix

Erbidium committed:
Remove unused files

Erbidium committed:
Fix time conversion


Fix

Erbidium committed:
Remove helper


Remove provide

Erbidium committed:
Fix dates comparison


Fix ranges merge and comparison

Erbidium committed:
Create helpers and separate functions

Erbidium committed:
Create helpers

Erbidium committed:
Create moment helpers


Lint fix

Erbidium committed:
Add time helper

Erbidium committed:
Remove last migration

Erbidium committed:
Merge branch 'development' into bug/exclusion-dates-timezones-and-month-change

Erbidium committed:
Fix styles, fix snapshot

Erbidium committed:
Add removed migration

Erbidium committed:
Reset styles

Erbidium committed:
Fix removed line

Erbidium committed:
Merge remote-tracking branch 'origin/development' into bug/exclusion-dates-timezones-and-month-change

Erbidium committed:
Fix dates merge bugs

Erbidium committed:
Merge branch 'development' into bug/exclusion-dates-timezones-and-month-change
